src/serac/physics/tests/solid.cpp

```diff
@@ -100,6 +100,177 @@ void functional_solid_test_static_J2()
   // EXPECT_LT(norm(solid_solver.reactions()), 1.0e-5);
 }
 
+// The purpose of this test is to check that the spatial function-defined essential boundary conditions are
+// working appropriately. It takes a 4 hex cube mesh and pins it in one corner. The z-direction displacement
+// is set to zero on the bottom face and a constant negative value on the top face.
+void functional_solid_spatial_essential_bc()
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  constexpr int p                   = 1;
+  constexpr int dim                 = 3;
+  int           serial_refinement   = 1;
+  int           parallel_refinement = 0;
+
+  // Create DataStore
+  axom::sidre::DataStore datastore;
+  serac::StateManager::initialize(datastore, "solid_mechanics_spatial_essential");
+
+  // Construct the appropriate dimension mesh and give it to the data store
+  std::string filename = SERAC_REPO_DIR "/data/meshes/onehex.mesh";
+
+  auto mesh = mesh::refineAndDistribute(buildMeshFromFile(filename), serial_refinement, parallel_refinement);
+  serac::StateManager::setMesh(std::move(mesh));
+
+  // Construct a functional-based solid mechanics solver
+  SolidMechanics<p, dim> solid_solver(
+      solid_mechanics::default_nonlinear_options, solid_mechanics::default_linear_options,
+      solid_mechanics::default_quasistatic_options, GeometricNonlinearities::On, "solid_mechanics");
+
+  solid_mechanics::LinearIsotropic mat{1.0, 1.0, 1.0};
+  solid_solver.setMaterial(mat);
+
+  // Set up
+  auto zero_vector   = [](const mfem::Vector&, mfem::Vector& u) { u = 0.0; };
+  auto zero_scalar   = [](const mfem::Vector&) { return 0.0; };
+  auto scalar_offset = [](const mfem::Vector&) { return -0.1; };
+
+  auto is_on_bottom = [](const mfem::Vector& x) {
+    if (x(2) < 0.01) {
+      return true;
+    }
+    return false;
+  };
+
+  auto is_on_bottom_corner = [](const mfem::Vector& x) {
+    if (x(0) < 0.01 && x(1) < 0.01 && x(2) < 0.01) {
+      return true;
+    }
+    return false;
+  };
+
+  auto is_on_top = [](const mfem::Vector& x) {
+    if (x(2) > 0.95) {
+      return true;
+    }
+    return false;
+  };
+
+  solid_solver.setDisplacementBCs(is_on_bottom_corner, zero_vector);
+  solid_solver.setDisplacementBCs(is_on_bottom, zero_scalar, 2);
+  solid_solver.setDisplacementBCs(is_on_top, scalar_offset, 2);
+
+  // Set a zero initial guess
+  solid_solver.setDisplacement(zero_vector);
+
+  // Finalize the data structures
+  solid_solver.completeSetup();
+
+  // Perform the quasi-static solve
+  double dt = 1.0;
+  solid_solver.advanceTimestep(dt);
+  solid_solver.outputState();
+
+  auto [size, rank] = serac::getMPIInfo();
+
+  mfem::Vector rank_0_displacement(54);
+  rank_0_displacement(0)  = -1.0617852403956275e-15;
```

Comment:
:dizzy_face:

Reply:
I honestly just copy/pasted from the computed solution. I can go back through and clean this up.